speechbrain/core.py

```diff
@@ -1050,60 +1068,89 @@ def fit_batch(self, batch):
         -------
         detached loss
         """
-        valid_loss = False
-
-        # Managing automatic mixed precision
-        if self.auto_mix_prec:
-            with torch.autocast(device_type=torch.device(self.device).type):
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-
-            # Losses are excluded from mixed precision to avoid instabilities
-            loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    self.scaler.scale(
-                        loss / self.grad_accumulation_factor
-                    ).backward()
-                if should_step:
-                    self.scaler.unscale_(self.optimizer)
-                    self.scaler.step(self.optimizer)
-                    self.scaler.update()
-                    self.zero_grad()
-                    self.optimizer_step += 1
-        else:
-            if self.bfloat16_mix_prec:
-                with torch.autocast(
-                    device_type=torch.device(self.device).type,
-                    dtype=torch.bfloat16,
-                ):
-                    outputs = self.compute_forward(batch, Stage.TRAIN)
-                    loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-            else:
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-                loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
-                if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
+        amp = AutocastConfig.from_name(self.autocast)
+        should_step = (self.step % self.grad_accumulation_factor) == 0
+
+        # We wrap the forward pass in no_sync to avoid
+        # unnecessary synchronization of gradients.
+        # Empirically, we found that this improves performance by a lot (2x speedup).
```

Comment:
I'm not sure about this, according to https://discuss.pytorch.org/t/whats-no-sync-exactly-do-in-ddp/170259, forward pass and backward pass should be in the same `with self.no_sync` block to correctly skip gradient synchronization

Reply:
Yes, I do know. But as pointed out in the comments, some DDP models (e.g. HF w2v2) are slower when the `self.no_sync(not should_step)` is wrapper around the fwd. There's a real difference between `self.no_sync(not should_step)` and `self.no_sync()` around the fwd.

Reply:
I double checked using https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech/ASR/transducer, according to my tests this code leads to wrong results as it never synchronizes the gradients (gradient values should be identical in each rank every `grad_accumulation_factor` steps):

```python
with self.no_sync():
    with torch.autocast(
        dtype=amp.dtype, device_type=torch.device(self.device).type,
    ):
        outputs = self.compute_forward(batch, sb.Stage.TRAIN)

with self.no_sync(not should_step):
    # The objectives are removed from Autocast to avoid
    # potential numerical instabilities with CTC loss, etc.
    loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)

    scaled_loss = self.scaler.scale(
        loss / self.grad_accumulation_factor
    )
    scaled_loss.backward()

for p in self.modules.parameters():
    print("RANK: " + os.environ["RANK"], "STEP: " + str(self.step), "GRAD: " + str(p.grad.flatten()[0:5]))
    break

if should_step:
    self.optimizers_step()
```

```bash
python -m torch.distributed.launch --nproc_per_node 2 \
train.py hparams/conformer_transducer.yaml --data_folder LibriSpeech \
--grad_accumulation_factor 4 --d_model 128 --use_torchaudio True \
--distributed_launch --find_unused_parameters
```

```
...
RANK: 0 STEP: 1 GRAD: tensor([-0.2559, -0.2251, -0.1947, -0.1615, -0.1247], device='cuda:0')
RANK: 1 STEP: 1 GRAD: tensor([-0.1381, -0.1090, -0.0870, -0.1038, -0.0626], device='cuda:1')
RANK: 0 STEP: 2 GRAD: tensor([-0.5477, -0.4795, -0.3994, -0.3519, -0.27181, device='cuda:0')
RANK: 1 STEP: 2 GRAD: tensor([-0.5832, -0.5016, -0.4143, -0.3648, -0.2632], device='cuda:1')
RANK: 0 STEP: 3 GRAD: tensor([-0.7123, -0.6151, -0.5048, -0.4761, -0.3532], device='cuda:0')
RANK: 1 STEP: 3 GRAD: tensor([-0.8210, -0.6935, -0.5572, -0.5350, -0.3820], device='cuda:1')
--> RANK: 0 STEP: 4 GRAD: tensor([-0.7952, -0.6520, -0.5222, -0.5171, -0.3215], device='cuda:0")
--> RANK: 1 STEP: 4 GRAD: tensor([-0.8469, -0.6987, -0.5467, -0.5705, -0.3754], device='cuda:1')
```

```python
with self.no_sync(not should_step):
    with torch.autocast(
        dtype=amp.dtype, device_type=torch.device(self.device).type,
    ):
        outputs = self.compute_forward(batch, sb.Stage.TRAIN)

    # The objectives are removed from Autocast to avoid
    # potential numerical instabilities with CTC loss, etc.
    loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)

    scaled_loss = self.scaler.scale(
        loss / self.grad_accumulation_factor
    )
    scaled_loss.backward()

for p in self.modules.parameters():
    print("RANK: " + os.environ["RANK"], "STEP: " + str(self.step), "GRAD: " + str(p.grad.flatten()[0:5]))
    break

if should_step:
    self.optimizers_step()
```

```bash
python -m torch.distributed.launch --nproc_per_node 2 \
train.py hparams/conformer_transducer.yaml --data_folder LibriSpeech \
--grad_accumulation_factor 4 --d_model 128 --use_torchaudio True \
--distributed_launch --find_unused_parameters
```

```
...
RANK: 0 STEP: 1 GRAD: tensor([-0.2559, -0.2251, -0.1947, -0.1615, -0.1247], device='cuda:0')
RANK: 1 STEP: 1 GRAD: tensor([-0.1381, -0.1090, -0.0870, -0.1038, -0.0626], device='cuda:1')
RANK: 0 STEP: 2 GRAD: tensor([-0.5477, -0.4795, -0.3994, -0.3519, -0.27181, device='cuda:0')
RANK: 1 STEP: 2 GRAD: tensor([-0.5832, -0.5016, -0.4143, -0.3648, -0.2632], device='cuda:1')
RANK: 0 STEP: 3 GRAD: tensor([-0.7123, -0.6151, -0.5048, -0.4761, -0.3532], device='cuda:0')
RANK: 1 STEP: 3 GRAD: tensor([-0.8210, -0.6935, -0.5572, -0.5350, -0.3820], device='cuda:1')
--> RANK: 0 STEP: 4 GRAD: tensor([-0.8211, -0.6754, -0.5344, -0.5438, -0.3485], device='cuda:0')
--> RANK: 1 STEP: 4 GRAD: tensor([-0.8211, -0.6754, -0.5344, -0.5438, -0.3485], device='cuda:1')
```

Reply:
We should further investigate why we have a slowdown with w2v2, do you have a script/recipe to reproduce it?

Reply:
Thanks Luca for the details. I was not aware of this behaviour w.r.t gradients. 

Yes, you just have to run the w2v CommonVoice CTC recipe with and without the no_sync() around the fwd. You need to use DDP ofc (i.e., more than 2gpus).